### DISCUSSION TITLE
timeouts in protocol detection on http idle connections

### DISCUSSION THREAD

**Discussion by cHiv0rz** [linkerd/linkerd2#13116] [Help]
Hi all, I'm trying to understand a rare case I spotted out on my logs.

I have an EKS cluster, running a Linkerd edge-24.5.5 (I think it's independent of the version, but who knows). Checking the logs, I noticed we have several timeouts on protocol detection between two of our golang services, connecting via HTTP. Apparently this timeouts must affect to the latency, but is not the case, which is weird.

After a couple of testing session, I found that the "issue" was related to idle connections. For some reasons we have set a timeout of 90s on the http client, and it seems this timeout is related to that.

If I'm not mistaken, the protocol detection reads a few bytes of the traffic to detect what protocol is using, but I couldn't find in your documentation what is the size necessary for it, and if it could be related to the keep alize size is not enough for detect protocol.

Thanks

**Comment by alpeb** [comment]
> If I'm not mistaken, the protocol detection reads a few bytes of the traffic to detect what protocol is using, but I couldn't find in your documentation what is the size necessary for it, and if it could be related to the keep alize size is not enough for detect protocol.

Right, if the client is just opening a connection and not sending data, then the proxy won't be able to detect the protocol. As for keep-alives, are those TCP keep-alives? If I'm not mistaking, those aren't enough and you need to send some actual payload at the application layer for the protocol to be detected.

**Reply by cHiv0rz** [reply]
That's what I thought, thanks :)
